### ISSUE TITLE
Question about prometheus metrics socket calculation

### ISSUE THREAD

**Issue by chenxinlong** [vernemq/vernemq#1466]
### Environment

- VerneMQ Version: 1.7.1
- OS: CentOS 7
- Erlang/OTP version (if building from source): n/a
- Cluster size/standalone: cluster size 1

### Expected behavior

The metrics of socket_open - socket_close equals to the output of the netstat command 

### Actual behavior

As the pic below, the sum of vernemq listeners socket num is about 2700. But the metrics of socket_open and socket_close is 13834 and 8017 which means the opening vernemq socket num is about 5800.

![image](https://user-images.githubusercontent.com/9261954/78564925-ba34e680-784f-11ea-86c5-829fcfb37229.png)

**Comment by ioolkos** [comment]
@chenxinlong thanks & sorry, I don't have an explanation for this.

**Comment by vytasmk** [comment]
Hello, looking at the exact same situation on v1.9.2 that is running on official Docker image.

I was looking at the Grafana graphs and did not understood what is happening and why open sockets are growing faster than closed sockets. See pic. below where "Connected clients" are calculated from `socket_open - socket_closed` counters. After fresh restart broker counters were showing correct results and graph was showing active client count but after 14 days client count started growing (added another graph).

![image](https://user-images.githubusercontent.com/37620896/83545796-63beee00-a508-11ea-81d2-167f621322f1.png)

Have checked opened connections with `netstat` and it shows correct values
```
netstat -an | grep ESTABLISHED | wc -l
6345
```

Current metrics looks like this:
```
# HELP socket_open The number of times an MQTT socket has been opened.
# TYPE socket_open counter
socket_open{node="VerneMQ@172.18.0.4"} 4241486
# HELP socket_close The number of times an MQTT socket has been closed.
# TYPE socket_close counter
socket_close{node="VerneMQ@172.18.0.4"} 4025181
# HELP socket_close_timeout The number of times VerneMQ closed an MQTT socket due to no CONNECT frame has been received on time.
# TYPE socket_close_timeout counter
socket_close_timeout{node="VerneMQ@172.18.0.4"} 286803
# HELP socket_error The total number of socket errors that have occurred.
# TYPE socket_error counter
socket_error{node="VerneMQ@172.18.0.4"} 27
```

As I understand there is no other metrics (gauge) to show active clients?

**Comment by ioolkos** [comment]
@vytasmk can you post full output of `vmq-admin metrics show`?

anything from the logs when this started?

**Comment by vytasmk** [comment]
```
bash-4.4$ vmq-admin metrics show
counter.queue_initialized_from_storage = 0
counter.queue_message_expired = 0
counter.mqtt_auth_sent = 0
counter.queue_message_in = 286871242
counter.queue_setup = 3699669
gauge.retain_memory = 21184
gauge.system_utilization_scheduler_9 = 0
counter.mqtt_connack_accepted_sent = 0
counter.mqtt_subscribe_error = 0
counter.mqtt_suback_sent = 4360918
counter.mqtt_connack_sent = 3722377
counter.mqtt_puback_invalid_error = 0
gauge.vm_memory_atom = 2033945
counter.mqtt_publish_sent = 275624569
counter.netsplit_resolved = 0
counter.bytes_received = 413465873663
counter.mqtt_pingreq_received = 433882620
counter.mqtt_pubrec_invalid_error = 0
gauge.system_run_queue = 1
counter.netsplit_detected = 0
gauge.queue_processes = 6325
counter.mqtt_unsubscribe_error = 0
counter.mqtt_pubcomp_sent = 238079721
gauge.system_utilization_scheduler_3 = 35
counter.system_reductions = 1819884705585
counter.mqtt_connect_received = 3934750
gauge.vm_memory_system = 464896432
gauge.system_utilization_scheduler_13 = 0
counter.cluster_bytes_received = 0
counter.mqtt_auth_received = 0
counter.socket_close_timeout = 287291
counter.mqtt_pubrec_sent = 241048453
gauge.system_utilization_scheduler_10 = 0
counter.client_keepalive_expired = 337484
gauge.system_utilization_scheduler_4 = 2
counter.system_runtime = 616507256
counter.router_matches_remote = 0
counter.system_wallclock = 2160149042
counter.queue_message_unhandled = 252950
counter.system_context_switches = 14626937000
gauge.system_utilization_scheduler_7 = 2
counter.mqtt_connack_unacceptable_protocol_sent = 0
counter.mqtt_pubrec_received = 0
counter.socket_error = 27
counter.mqtt_pubrel_sent = 0
counter.socket_close = 4039028
counter.mqtt_publish_auth_error = 9849
counter.mqtt_invalid_msg_size_error = 0
counter.cluster_bytes_dropped = 0
counter.mqtt_connack_not_authorized_sent = 0
gauge.vm_memory_atom_used = 2012143
counter.mqtt_publish_error = 0
counter.mqtt_pingresp_sent = 433882620
gauge.system_utilization_scheduler_6 = 3
gauge.system_utilization_scheduler_14 = 0
counter.mqtt_pubrel_received = 238080544
counter.router_matches_local = 286920555
counter.system_exact_reductions = 1819716683852
counter.queue_message_drop = 31743
counter.mqtt_connack_bad_credentials_sent = 0
counter.bytes_sent = 444129075035
gauge.system_process_count = 25541
counter.mqtt_pubcomp_received = 0
counter.mqtt_puback_received = 0
gauge.system_utilization_scheduler_5 = 3
gauge.system_utilization_scheduler_11 = 0
counter.system_words_reclaimed_by_gc = 2952502558522
gauge.router_subscriptions = 118403
counter.mqtt_connack_server_unavailable_sent = 0
gauge.system_utilization_scheduler_12 = 0
gauge.system_utilization_scheduler_1 = 2
counter.client_expired = 0
counter.mqtt_disconnect_received = 2152533
gauge.vm_memory_binary = 77027152
counter.mqtt_puback_sent = 36002
counter.system_io_in = 456539459711
counter.mqtt_unsuback_sent = 1000124
counter.mqtt_connack_identifier_rejected_sent = 0
counter.mqtt_subscribe_received = 4363955
counter.mqtt_subscribe_auth_error = 3037
counter.mqtt_disconnect_sent = 0
gauge.vm_memory_code = 13831773
counter.system_io_out = 637523096639
counter.mqtt_unsubscribe_received = 1000124
counter.system_gc_count = 4539976227
gauge.system_utilization_scheduler_15 = 0
gauge.system_utilization = 9
gauge.vm_memory_processes_used = 297655136
gauge.system_utilization_scheduler_16 = 0
counter.queue_message_out = 275624569
gauge.vm_memory_total = 762574696
gauge.vm_memory_processes = 297678264
gauge.router_memory = 98770536
counter.mqtt_pubcomp_invalid_error = 0
gauge.retain_messages = 0
gauge.system_utilization_scheduler_8 = 1
counter.queue_teardown = 3693344
counter.cluster_bytes_sent = 0
gauge.vm_memory_ets = 345566216
counter.socket_open = 4257722
gauge.system_utilization_scheduler_2 = 100
counter.mqtt_publish_received = 272930951
```

There are usually around 6100 active clients that come from all ower the world and stay connected. They send some statistics as JSON every 2 min. so not so much traffic. If they disconnect they try to reconnect immediately but needs to connect and disconnect three times (such logic was created few years ago): first connection nonSSL to retrieve CA scertificate, then disconnect and reconnect to SSL port as anonymous client to retrieve username and password, and third time disconnects and connects to SSL port with provided credentials. So counters increases quite fast. But I think that should not be the problem as two weeks there was no problems.

In the logs I can not tell anything as they are rotated also I did not expected such behavior and I have only spotted this anomaly today. I plan to upgrade system where this broker is running so will check then every day until it starts again. I have another instance running as a beta system with less clients (~60) but it is running only few days now after reboot. So will keep an eye on it to. also it is not clear does it depends on time or on event count.

**Comment by ioolkos** [comment]
Hm, interesting. 
There's a high number of `counter.socket_close_timeout = 287291` but other than that I can't spot anything unusual for the moment.
Could some system resources be silently exhausted? File descriptors or the like (OTOH this would certainly show in the logs).
The 3x connection "handshake" your clients are doing is also interesting.

**Comment by ilpanich** [comment]
Hi, seen this issue on VerneMQ 1.12.3... Any news about it?

**Comment by ioolkos** [comment]
@ilpanich Thanks, I think what you mention is fixed by PR https://github.com/vernemq/vernemq/pull/2000

Note that the closed socket stats exported to the metrics come from elsewhere. This is not about the Webhooks cache stats.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ilpanich** [comment]
Nice!
Thank you @ioolkos. Do you have any schedule for this fix release?

We have seen in the code that that the closed socket stats are computed elsewhere. What is quite strange in our environment is that on at least two different brokers the number of errors related to bug fixed in PR #2000 is compatible with the difference between the number of active queues and the number of active sockets metrics.
